[WPE][WebDriver] Support taking the screenshots from the UIProcess

lauromoura opened the pull request:
#### 50c1d97fdef710ce465805de08a4d2d6c3c788f4
<pre>
[WPE][WebDriver] Support taking the screenshots from the UIProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=299732">https://bugs.webkit.org/show_bug.cgi?id=299732</a>

Reviewed by Carlos Garcia Campos.

Currently, WPE&apos;s implentation of taking screenshots rely on asking the
root `RenderLayer` on the WebProcess to paint its content. While it
worked fine for most elements, this approach could lead to differences
from the final composited frame that we send back to the UIProcess,
especially for some kinds of accelerated content, like animated 3D CSS
filters.

On top of that, the recent ANGLE bump in 295416@main included a stricter
check on ReadPixels that affected taking screenshots of WebGL scenes on
WebGL 1.0 environments, as the GraphicsContextGLAngle used it to read
the WebGL scenes contents for this code path.

To address these issues, this commits adds support for taking
screenshots on the UIProcess, by reading the last buffer committed to
the system.

* Source/WebKit/Shared/WebBackForwardListItem.h:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::takeViewSnapshot):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::takeViewSnapshot):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::takeScreenshot):
* Source/WebKit/UIProcess/Automation/skia/WebAutomationSessionSkia.cpp:
(WebKit::base64EncodedPNGData):
(WebKit::WebAutomationSession::platformGetBase64EncodedPNGData):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/ViewSnapshotStore.cpp:
* Source/WebKit/UIProcess/ViewSnapshotStore.h:
(WebKit::ViewSnapshot::image const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/skia/ViewSnapshotSkia.cpp: Copied from Source/WebKit/UIProcess/Automation/skia/WebAutomationSessionSkia.cpp.
(WebKit::ViewSnapshot::create):
(WebKit::ViewSnapshot::ViewSnapshot):
(WebKit::ViewSnapshot::hasImage const):
(WebKit::ViewSnapshot::clearImage):
(WebKit::ViewSnapshot::estimatedImageSizeInBytes const):
(WebKit::ViewSnapshot::size const):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp:
(WebKit::getImageInfoFromBuffer):
(WebKit::saveBufferSnapshot):
(WebKit::AcceleratedBackingStore::takeSnapshot):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.h:
* WebDriverTests/TestExpectations.json:

Canonical link: <a href="https://commits.webkit.org/301948@main">https://commits.webkit.org/301948@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/965e9ff66aada9b922e56a505e0beb663fdbd960

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126748 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37359 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133705 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78381 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47020 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54927 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96451 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64496 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129696 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37622 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113365 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76975 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36503 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77098 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107441 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31848 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136282 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53426 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41112 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104964 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53920 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109725 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104667 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26841 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50164 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28500 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50879 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53354 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59152 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52613 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55951 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54362 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/69dceed04a39ac8ca838899490d9113d1c9d27f2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51535

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7d9bd90352ed571885b17e2d4c9e0bde7066cd16. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51535

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
This should be endif, we want to fail the same way when building with the new api but the old one is used.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
```suggestion
```

carlosgcampos commented on Source/WebKit/UIProcess/Automation/skia/WebAutomationSessionSkia.cpp:
You can use SkData::bytesSpan()

lauromoura replied:
Good catch. I'll also refactor this funtion to use the `base64EncodedPNGData(SkImage&)` defined above and use `byteSpan` there.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
I don't think we need this async version, the previous call to callAfterNextPresentationUpdate should ensure we have a pending buffer ready, I think.

carlosgcampos commented on Source/WebKit/UIProcess/skia/ViewSnapshotSkia.cpp:
Don't we have to call `ViewSnapshotStore::singleton().didAddImageToSnapshot(*this)` ?

carlosgcampos commented on Source/WebKit/UIProcess/skia/ViewSnapshotSkia.cpp:
And `ViewSnapshotStore::singleton().willRemoveImageFromSnapshot(*this)` here?

carlosgcampos commented on Source/WebKit/UIProcess/skia/ViewSnapshotSkia.cpp:
We could probably use `m_image->imageInfo().computeMinByteSize()`

carlosgcampos commented on Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp:
I think we should try to not use this async approach and get the snapshot after the call after next presenation update using the pending buffer if there's any or committed buffer otherwise.

carlosgcampos commented on Source/WebKit/UIProcess/ViewSnapshotStore.cpp:
Why?

lauromoura replied:
I removed it because it relied on taking the screenshots synchronously. I'll test this approach, reading the last `m_commitedBuffer`, so we can avoid this guard.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6b86043e0f2012d53e469e6c4a52a3de7771587d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51535

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/284ef69fe2080ae238bed12b3f665434779f95a5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51535

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4f4724867f37cbe09714ac82792bf27235cbb929. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51535

lauromoura commented:
@carlosgcampos I'm afraid the `#if` checks still seem to be a bit confusing.

Maybe `ViewSnapshotStoreGtk3.cpp` could be renamed to `ViewSnapshotCairo.cpp`, alongside adding some `ENABLE(VIEWSNAPSHOT_SKIA)` and `ENABLE(VIEWSNAPSHOT_CAIRO)` helpers? WDYT?

lauromoura commented on Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp:
```diff
@@ -155,6 +167,76 @@ void AcceleratedBackingStore::frame(uint64_t bufferID, Rects&& damageRects, WTF:
+Expected<Ref<ViewSnapshot>, String> AcceleratedBackingStore::takeSnapshot(std::optional<WebCore::IntRect>&& clipRect)
+{
+    if (!m_committedBuffer) [[unlikely]]
+        return makeUnexpected("No commited buffer to create snapshot from"_s);
```
Done. But should we do any kind of check regarding the fences to read from the pending buffer?

lauromoura commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13648,6 +13644,29 @@ RefPtr<ViewSnapshot> WebPageProxy::takeViewSnapshot(std::optional<WebCore::IntRe
+    if (!pageClient)
+        return makeUnexpected("No PageClient available to handle snapshot request"_s);
+
+    return pageClient->takeViewSnapshotWithErrorDetail(WTFMove(clipRect));
```
As mentioned in the previous reply, this was postponed to a follow-up commit, in https//webkit.org/b/300271

lauromoura committed:
[WPE][WebDriver] Support taking the screenshots from the UIProcess
https://bugs.webkit.org/show_bug.cgi?id=299732

Reviewed by Carlos Garcia Campos.

Currently, WPE's implentation of taking screenshots rely on asking the
root `RenderLayer` on the WebProcess to paint its content. While it
worked fine for most elements, this approach could lead to differences
from the final composited frame that we send back to the UIProcess,
especially for some kinds of accelerated content, like animated 3D CSS
filters.

On top of that, the recent ANGLE bump in 295416@main included a stricter
check on ReadPixels that affected taking screenshots of WebGL scenes on
WebGL 1.0 environments, as the GraphicsContextGLAngle used it to read
the WebGL scenes contents for this code path.

To address these issues, this commits adds support for taking
screenshots on the UIProcess, by reading the last buffer committed to
the system.

* Source/WebKit/Shared/WebBackForwardListItem.h:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::takeViewSnapshot):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::takeViewSnapshot):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::takeScreenshot):
* Source/WebKit/UIProcess/Automation/skia/WebAutomationSessionSkia.cpp:
(WebKit::base64EncodedPNGData):
(WebKit::WebAutomationSession::platformGetBase64EncodedPNGData):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/ViewSnapshotStore.cpp:
* Source/WebKit/UIProcess/ViewSnapshotStore.h:
(WebKit::ViewSnapshot::image const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/skia/ViewSnapshotSkia.cpp: Copied from Source/WebKit/UIProcess/Automation/skia/WebAutomationSessionSkia.cpp.
(WebKit::ViewSnapshot::create):
(WebKit::ViewSnapshot::ViewSnapshot):
(WebKit::ViewSnapshot::hasImage const):
(WebKit::ViewSnapshot::clearImage):
(WebKit::ViewSnapshot::estimatedImageSizeInBytes const):
(WebKit::ViewSnapshot::size const):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp:
(WebKit::getImageInfoFromBuffer):
(WebKit::saveBufferSnapshot):
(WebKit::AcceleratedBackingStore::takeSnapshot):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.h:
* WebDriverTests/TestExpectations.json:

Canonical link: https://commits.webkit.org/301948@main
